File: cs/app/arraychunk.cs
// given an array and chunk size.  divide the array into many sub arrays
// where each subarray is of length size
// eg: chunk([1,2,3,4],2) => [[1,2],[3,4]]
// eg: chunk([1,2,3,4,5],2) => [[1,2],[3,4], [5]]
// eg: chunk([1,2,3,4,5, 6, 7, 8],3) => [[1,2, 3],[4,5,6], [7,8]]
using System;
using System.Collections.Generic;
using System.Linq;

public static class ArrayChunk{


    public static List<int[]> ChunkArray(int[] arr, int size)
    {
        if (arr == null || size <= 0)
        {
            throw new ArgumentException("Invalid input");
        }

        //list of integer arrays eg: [1,2], [3,4] ,etc.
        List<int[]> result = new();

        for (int i = 0; i < arr.Length; i += size)
        {
            //go one by one, incrementing by size.
            //use take method to grab X integers and add to list via "Toarray"
            //eg: skip 2, get 3 and 4 (2 if possible), conver to array, add to result
            result.Add(arr.Skip(i).Take(size).ToArray());
        }

        return result;
    }

    public static List<int[]> ChunkArray2(int[] arr, int size)
    {
        List<int[]> chunked = new List<int[]>();
        List<int> chunk = new List<int>();
        int count = 0;
        foreach (int i in arr)
        {

            chunk.Add(i);
            if (chunk.Count == size) //array is at chunk size, add
            {
                chunked.Add(chunk.ToArray()); //adding array not list
                chunk.Clear();
            }
            count++;
            if (count == arr.Length) //reached last element
            {
                chunked.Add(chunk.ToArray());
                chunk.Clear();
            }
        }
        return chunked;
    }
}

File: js/arraychunk.js
// given an array and chunk size.  divide the array into many sub arrays
// where each subarray is of length size
// eg: chunk([1,2,3,4],2) => [[1,2],[3,4]]
// eg: chunk([1,2,3,4,5],2) => [[1,2],[3,4], [5]]
// eg: chunk([1,2,3,4,5, 6, 7, 8],3) => [[1,2, 3],[4,5,6], [7,8]]


function test1(chunkArr, size) {

    let outer = [];
    let inner = [];
    let output = [];
    let totaloutput = [];
    let count = 0;

    for (let outer of chunkArr)
    {
        for (let inner in outer) //of?
        {
            output[count] = inner;
            count++;
            if (count == size)
            {
                totaloutput.push(output);
                output = [];
                count = 0;
            }
        }      
    }

    return totaloutput;

     
}

function test2(array, size) {
    
    const chunked = [];

    for (let element of array) {
        
        const last = chunked[chunked.length-1];
      
        if (!last || last.length === size){ //1 and 3
            
            //put new chunk in and set. creates new array and puts
            //element into it as well. 
            chunked.push([element]);         
            
        }else{ //2 and 4
            last.push(element); //confusing. basically pushes to chunked as well. 
            //i guess pointer
            
        }        
    }
    return chunked;
}

function test3(array, size) {
    const chunked  = [];

    let index = 0;

    while(index < array.length)
    {
        chunked.push(array.slice(index, index+size));

        index = index + size;
    }
    return chunked;
}

function test4(array, size) {
    //iterate over array
    //grab size
    //add to chunk
    let chunked = [];
    let chunk = [];
    let ctr = 0;
    let ctr2 = 0;
    for (let char of array) {
        chunk.push(char);
        ctr++; ctr2++;
        if (ctr == size || ctr2 == array.length) {
            chunked.push(chunk);
            chunk = [];
            ctr = 0;
        }
    }
    return chunked;
}
//console.log(test1([1,2,3,4],2));
console.log(test2([1,2,3,4],2));
console.log(test2([1,2,3,4,5],2));
console.log(test2([1,2,3,4,5, 6, 7, 8],3));

console.log(test3([1,2,3,4],2));
console.log(test3([1,2,3,4,5],2));
console.log(test3([1, 2, 3, 4, 5, 6, 7, 8], 3));

console.log(test4([1, 2, 3, 4, 5, 6, 7, 8], 3));



//console.log(test3('apple'));

module.exports = test1;